zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java

```diff
@@ -285,6 +298,20 @@ protected Object jobRun() throws Throwable {
       throw new RuntimeException("Can not find interpreter for " + getRequiredReplName());
     }
 
+    if (this.user != null &&
+      !factory.getInterpreterSettings(note.getId()).isEmpty()) {
+      for (InterpreterSetting intp: factory.getInterpreterSettings(note.getId())){
+
+        if (replName.startsWith(intp.getName()) &&
+          intp.getOption().isSetPermission() &&
+          !hasPermission(authenticationInfo.getUser(), intp.getOption().getUsers())) {
+          logger.error("{} has no permission for {} ", authenticationInfo.getUser(), repl);
```

Comment:
Just want to give you a perspective of somebody who reads this code: it is VERY hard to understand and reason what happens in both conditions here.

By looking at **first condition**:

```java
this.user != null && !factory.getInterpreterSettings(note.getId()).isEmpty()
```

I would assume that is some kind of pre-condition check, because we do not do anything in case it is `false`.
But if that is `true`, we do something for each `interpreter`. 

Then I would look at both conditions to understand, what they mean, starting from `this.user != null`. As a reader, I have no idea what is the case when `this.user` can be `null`. Is that because of some kind of error condition? Or is that because auth is not enabled? Or is that the case when user is not authorized? It's very hard to say, without spending a lot of time to understand all the cases

Same for `factory.getInterpreterSettings(note.getId()).isEmpty()` - what does it mean when interpreter settings are empty for a note? When can that happen? Why do we check for it here? Ah, after some thinking and looking for other code that uses it - I can see, that is some kind of collection!

And all this questions for reader could be avoided, and time saved, if we give a reader a clue in form of readable names for this conditions in english - we can extract this logic to the functions that self-describe it's purpose through the function names (you can think of it as a form of documentation):

```java
if (this.noteHasUser() && this.noteHasInterpreters()) {
  //...
}

boolean noteHasUser() {
  return this.user != null;
}

boolean noteHasInterpreters() {
  // BTW, this is used in 2 other places and looks like candidate for further simplification
  // return !getInterpreterListFor(this.note).isEmpty();
  return !factory.getInterpreyterSettings(note.getId()).isEmpty();
}
```

--------------------

By looking at **second condition**:

````java
replName.startsWith(intp.getName()) &&
 +          intp.getOption().isSetPermission() &&
 +          !hasPermission(authenticationInfo.getUser(), intp.getOption().getUsers())
```

as a reader, for me it's even harder to say, what case does this represent.

Judging by the loop and an error message - it looks this code does to things:
 - finds current interpreter by name
 - only for the current interpreter, it checks if user has access to it

So a function can be extracted here as well:

```java
boolean isUserAuthorizedToAccessInterpreter(String user,  InterpreterOption intpOpt)){
  return intpOpt.isSetPermission() &&
         !hasPermission(authenticationInfo.getUser(), intpOpt.getUsers())
}
```

But it can be even further simplified if we decouple logic for finding current `interp` in collection by `replName` to a new function `findInerpreterByName` (from the logic for checking if user is authorized to access the `interp`):

```java
InterpreterSettings findInerpreterByName(String name) {
  InterpreterSettings intp = null; // this would rather be `Option()` or `InterpreterSettings.NONE` or anything else type-safe
  for (InterpreterSetting i: factory.getInterpreterSettings(note.getId())) { // could re-use `getInterpreterListFor(this.note)` from above
    if (i.getName().startsWith(name)) {
      interp = i;
    }
 }
 return intp;
}
```

So, **final results**

*Before*

```java
if (this.user != null &&
  !factory.getInterpreterSettings(note.getId()).isEmpty()) {
  for (InterpreterSetting intp: factory.getInterpreterSettings(note.getId())){

    if (replName.startsWith(intp.getName()) &&
      intp.getOption().isSetPermission() &&
      !hasPermission(authenticationInfo.getUser(), intp.getOption().getUsers())) {
      return new InterpreterResult(...);
    }
 }
}
```

*After*

```java
if (this.noteHasUser() && this.noteHasInterpreters()) {
  InterpreterSetting intp = findInerpreterByName(replName);
  if (intp != null && 
      isUserAuthorizedToAccessInterpreter(authenticationInfo.getUser(), intp.getOption())) {
    return new InterpreterResult(...);
  }
}
```

I'm not native english speaker, but the resulting code very much looks like an english sentence. It explains the logic for the reader, who can follow the implementation details by looking at the new functions.

What do you think, does this make sense?